src/Http/Owin/src/OwinEnvironment.cs

```diff
@@ -254,7 +254,38 @@ public OwinEnvironment(HttpContext context)
 
     void ICollection<KeyValuePair<string, object>>.CopyTo(KeyValuePair<string, object>[] array, int arrayIndex)
     {
-        throw new NotImplementedException();
+        if (array is null)
+        {
+            throw new ArgumentNullException(nameof(array));
+        }
+        if (arrayIndex < 0)
+        {
+            throw new ArgumentOutOfRangeException(nameof(arrayIndex));
+        }
+        if (arrayIndex + _entries.Count + _context.Items.Count > array.Length)
+        {
+            throw new ArgumentException("Not enough available space in array", nameof(array));
+        }
+
+        // Simpler code and easier to read, but causes an allocation of the enumerator/iterator - is the trade off okay?
+        //foreach (var entryPair in this)
```

Comment:
Yes, go with the simpler code, it avoids redundancy and makes it easier to maintain.